### PR TITLE
Follow-up cleanups for model inputs

### PR DESCRIPTION
**Pull request by crisbeto**
Includes some cleanups around `model()` inputs as well as more test coverage. See the individual commits for more info.

### CODE REVIEW

File: packages/core/src/render3/instructions/listener.ts [hunk 1]
```diff
@@ -207,10 +207,16 @@ export function listenerInternal(
               directiveInstance.constructor.name}'.`);
         }
 
-        const subscription = (output as SubscribableOutput<unknown>).subscribe(listenerFn);
+        const subscriptionOrCallback =
+            (output as SubscribableOutput<unknown>).subscribe(listenerFn);
         const idx = lCleanup.length;
-        lCleanup.push(listenerFn, subscription);
-        tCleanup && tCleanup.push(eventName, tNode.index, idx, -(idx + 1));
+        lCleanup.push(listenerFn, subscriptionOrCallback);
```
Comment (crisbeto) [inline packages/core/src/render3/instructions/listener.ts:213]: We could simplify it a bit by tracking the `Subscription.unsubscribe` here, but then we run the risk of either the `this` not being correct when it's invoked, or having to track the context of the function, or having to create a new instance using `.bind`.

File: packages/core/src/render3/instructions/listener.ts [hunk 2]
```diff
@@ -277,8 +283,8 @@ function wrapListener(
 }
 
 /** Describes a subscribable output field value. */
-export interface SubscribableOutput<T> {
-  subscribe(listener: (v: T) => void): {unsubscribe: () => void;};
+interface SubscribableOutput<T> {
+  subscribe(listener: (v: T) => void): {unsubscribe: () => void;}|(() => void);
```
Comment (devversion) [inline packages/core/src/render3/instructions/listener.ts:287]: Nice! need this for output as well. Set this up as a preparation and to be able to "implement" the interface

File: packages/core/src/render3/node_manipulation.ts [hunk 3]
```diff
@@ -442,7 +442,7 @@ function processCleanups(tView: TView, lView: LView): void {
         const targetIdx = tCleanup[i + 3];
         ngDevMode && assertNumber(targetIdx, 'cleanup target must be a number');
         if (targetIdx >= 0) {
-          // unregister
+          // Destroy a QueryList or unsubscribe from a ModelSignal
```
Comment (devversion) [inline packages/core/src/render3/node_manipulation.ts:445]: Nit: Would probably mention, or anything callback-based destroy functions (need the same for output)

**Review by devversion — APPROVED** [review]
LGTM. Thx for all the tests! I realize I also should add one for `references_and_rename`. All others we have

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api, language-service

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 0b955b867c96024d2702234a47e5a0ef95e65b95.

**Pull request closed without merging**